### ISSUE TITLE
Two lines can be equals except for their type (eg. a valley and a moutain fold both having the same ends)

### ISSUE THREAD

**Issue by JinoBetti** [oripa/oripa#99]
According to several comments in the code, to lines having the same ends but not the same type is something that should be impossible to achieve (e.g. comment on `OriLine::equals` : line type is not compared because there is only one line in the real world if the two ends of the line are determined. )

While trying stuff out, I found out that having two lines that have the same ends but not the same type is possible, though. It's actually quite tricky to do it, but I remembered having done that while using the software before:
the steps to get to this situation are for example : 

1. Show the 4*4 grid
2. Draw a moutain line at `(0,0,200,0)`
2. Draw a valley line at `(0,0,100,0)`, this is possible because `removeDuplicationsFromInputLines` is called before the splitting process of the new line, so we can have two lines on one another, which I think is the biggest problem in this issue)
3. Draw a valley line at `(100,0,100,100)`, this will cause the first moutain line to be split into `(0,0,100,0)` and `(0,0,100,200)`

=> We have both a Moutain line and a Valley line at `(0,0,100,0)`

**Conclusions** : 
- I think it would be better to also test for equality of the line type in the equals method (even though it does not change anything for this bug)
- I think that the dupplication removal for the newly added lines should be done after splitting the current lines rather than before, (or maybe both?)... I'm not sure yet of the right course of action here

**Comment by Ooouch1** [comment]
On ver. 1.45, the valley line (0,0,100,0) disappears after the instruction. Maybe it is because one of the duplicated lines is rejected by `LineManager`. Anyway, this is a bug.

Testing line type in the test of line equality is acceptable. I think it won't matter.

I think we can use `OverlappingLineExtractor.extract(Collection<OriLine>)` to detect duplication efficiently. This method can find duplications among given lines faster than a naive way.

Whether rejecting duplications on each input or not is a difficult issue because old ORIPA does not care duplication at input. If we reject duplication every time, we should show some message when rejection happens.

**Comment by JinoBetti** [comment]
This time I didn't jump right in the implementation, and i tried to think of all the cases to see what's best to do, here are my thoughts : 
![image](https://user-images.githubusercontent.com/7082695/115839390-e4d1ed00-a41a-11eb-95c5-59f1481f08ca.png)

When we input a segment that's not on the same line, nothing happens differently (I wrote the cases just to remember to add the unit test associated)

When we input a segment on the same support line, three main cases appear, and when there is an overlap, my idea would be to split the lines and to keep the newly added one (so if you input (-100,0,100,0, VALLEY) on an existing (0,0,200,0, MOUNTAIN), you wind up with (-100,0,0,0,VALLEY), (0,0,100,0,VALLEY) and (100,0,200,0, MOUNTAIN) for example)

What do you think of it?

**Comment by Ooouch1** [comment]
> when there is an overlap, my idea would be to split the lines and to keep the newly added one

Sounds useful.

Assuming that there are few overlaps, naive algorithm seems OK from the view point of computation time: label all segments by whether newly added or not, run the current `addAll()`, extract overlapping segments by `OverlappingLineExtractor.extract(allLines)`, make splits, and determine the type of the overlapping parts. This takes O(n^2) time.

Of course more sophisticated algorithm is welcome. Maybe we can improve the algorithm by putting overlapping segments into `Collection<Collection<OriLine>>` by modifying `extract(lines)`, which reduces the number of pairwise comparison to make splits.

**Comment by Ooouch1** [comment]
It may be nice to consider drawing a segment over connected segments with same direction, which also solves the case that there is one overlapping segment. For example, let's consider in 1D for simplicity, if we have (0,1, V), (1,2, M), (2,3, M), (3,4, M) and (2,5, V) is added then we will have (0,1, V), (1,2, M), (2,3, V), (3,4 V), (4, 5, V).

**Comment by JinoBetti** [comment]
Yes exactly 👍 

Since I'm trying to use TDD, I started this time by writing very simple code in a test class (in fact, in the `OverlappingLineExtractorTest` class) to detect overlap. I started without using the `OverlappingLineExtractor::extract` method at first, but by writing down my intents in very simple functions : 

First what I wanted to have, is a function that show the two conditions for the 2 lines to be considered as overlapping : first, being on the same support line (first line in the test cases I've written above), and second, not being in case one "no overlap"
`private boolean detectOverlap(OriLine l1, OriLine l2) { return areOnSameSupportLine(l1, l2) && !areDisjointSegments(l1, l2); }`

Then, I wrote both functions : 
- `areDisjointSegments` compares the x coordinates of `l1.p0`, `l1.p1`, `l2.p0` and `l2.p1` to check if both points of `l1` have lower x values than both points of `l2`
- `areOnSameSupportLine` calculates the value of the `y = ax+b` affine function corresponding to `l1` at x = `l2.p0.x` and at x = `l2.p1.x`, to check if their respective values are `l2.p0.y` (more or less epsilon) and `l2.p1.y` (more or less epsilon, which goes like that : 
`return abs(l1.getAffineYValueAt(l2.p0.x) - l2.p0.y) < EPS && abs(l1.getAffineYValueAt(l2.p1.x) - l2.p1.y) < EPS;`
=> if the affine functions test the same at two distinct points, then the a and b constants are the same for the two lines and the support line is the same. (There is a problem with this method for vertical lines so an extra security has been added : `if(l1.isVertical()) return l2.isVertical();`

Then, since all unit tests seems to pass, I tried to run it a million times to see if this method was faster or not that the `OverlappingLineExtractor::extract` method. The result seems to be a lot faster : 1000000 iterations in 23ms now vs 19541ms before (I ran the test a few times and always have similar results)

So my thoughts are : 
- Both computation times are ok since we never have one million lines to compare, but when this is the case, I usually go with the easiest solution to understand
- Is there something I might be missing here that the original method does and I'm not doing, that could take time to compute? at first glance, I'd say that maybe `hashFactory.create` could be costly(?)
- Do you agree with the method I used to check if both lines are on the same support line?
- Would you be ok if I used this method to implement that (provided that you agree with the method and i didn't forget anything that could explain the cost of the existing `extract` method)?
- One drawback I can see in my solution is that the small difference of angle that is allowed between the support lines of l1 and l2 to be considered the same depends on their lengths : 
![image](https://user-images.githubusercontent.com/7082695/115928072-d836ae80-a485-11eb-8016-441e8ca354fb.png)

**Comment by Ooouch1** [comment]
Comments:

* Your `detectOverlap(l1, l2)` seems to have the same responsibility of `OverlappingLineExtractor::isOverlap(line0, line1)`. The implementation of `detectOverlap(l1, l2)` is more straightforward.
* Does `areDisjointSegments(l1, l2)` consider the case when both l1 and l2 are vertical at the same x? I think you have to use y value in that case.
* How many segments are used for the speed test? `OverlappingLineExtractor::extract(lines)` is designed to have a good performance if the size of `lines` is large (like 10,000 which can happen for 100x100 grid) and the distribution of lines is uniform. `extract(lines)` finds all possible overlaps among given `lines`. Such a situation is more difficult than 1 vs. n. 1 vs. n version has been implemented as `extract(lines, target)`.

Answers to your questions:

>Is there something I might be missing here that the original method does and I'm not doing, that could take time to compute?

You seems to consider 1 segments vs. n segments overlap only. Be noticed that we have to implement `addAll(inputLines, currentLines)` efficiently for large size inputs, whose overlap detection can be O(n^2) (n segments vs. n segments). `addAll` is an important method because it is used in copy & paste etc.

> Do you agree with the method I used to check if both lines are on the same support line?

Basically yes although the approach is similar to using `AnalyticLine`.

> Would you be ok if I used this method to implement that

Yes but pay attention to efficient implementation of `addAll(inputLines, currentLines)` as I mentioned above.

> One drawback I can see in my solution is that the small difference of angle that is allowed between the support lines of l1 and l2 to be considered the same depends on their lengths :

If it gets a matter, we can rewrite to use `AnalyticLine`.

**Comment by JinoBetti** [comment]
Thank you for your answers, i'll make a pull request soon, I finished a version that I think is ok in terms of computation time and which passes all the tests so far (but I havent tried copy pasting and all yet)

**Comment by JinoBetti** [comment]
I'm considering switching to the `AnalyticLine` class instead of computing the support line equality another way, but I have a few questions:
- there is only an y axis intercept, which doesn't work well (as stated in the comments) when the line is vertical, I'm thinking of adding an x intercept as well, to be able to deal with this case, how do you feel about that?
- if the line happens to be a zero length line, the intercept (and the angle ?) value will give `NaN` values, how do you feel about throwing an exception when the length is < EPS ?

**Comment by Ooouch1** [comment]
> there is only an y axis intercept, which doesn't work well (as stated in the comments) when the line is vertical, I'm thinking of adding an x intercept as well, to be able to deal with this case, how do you feel about that?

If you will use `AnalyticLineHashFactory` and the intercept will be used for grouping only, I think it will work well to set x value to `intercept` field instead of setting `Double.MAX_VALUE` when the line is vertical. If you need x intercept for other use, you can add it.

> if the line happens to be a zero length line, the intercept (and the angle ?) value will give NaN values, how do you feel about throwing an exception when the length is < EPS ?

It would be OK. My design assumes zero length lines are removed for simplicity because geometric computation algorithm tends to be complicated and such exception checks can be annoying...

**Comment by Ooouch1** [comment]
@JinoBetti 
Hi, how is it going?
I hope you are doing well.

Here is a pseudo code of my idea:

* parameter: `Collection<OriLine> addedLines`, `Collection<OriLine> allLines`.
    * each line of addedLines should be split at cross points.
    * allLines is the result of adding lines and splitting at cross points by current `addAll()`. (So it includes all of `addedLines`.)
* By grouping overlaps among allLines by support line, we obtain `Collection<Collection<OriLine>> overlapGroups`.
  Such grouping can be done with a modification of `OverlappingLineExtractor::extract(lines)`.
* For each group in `overlapGroups`:
    * `sortedPoints` = all unique points of lines in the group, which are sorted by x (y if lines are vertical)
    * `addedOverlaps` = lines in both `addedLines` and the group
    * `existingOverlaps` = lines in the group and not in `addedOverlaps`
    * make `splitLines` by connecting the points in `sortedPoints` sequentially.
    * For each line in `splitLines`:
        * If the line overlaps a line of `addedOverlaps`, then the type of the line is the same as the overlapping line.
        * Otherwise, if the line ovrelaps a line of `existingOverlaps`, then the type of the line is the same as the overlapping line.
* remove all of `addedOverlaps` from allLines.
* remove all of `existingOverlaps` from allLines.
* add all of `splitLines` to allLines.

By using hash sets as `addedLines` and `allLines`, the computation time will be short.

**Comment by JinoBetti** [comment]
Hi ! Sorry for taking so long to answer, I've been really busy for a few weeks but I'll be able to resume working on this issue soon!
